apis/Google.Cloud.Firestore/Google.Cloud.Firestore/EnumHelpers.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright 2018, Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Reflection;
+using BclType = System.Type;
+
+namespace Google.Cloud.Firestore
+{
+    /// <summary>
+    /// Enum conversion helper methods.
+    /// </summary>
+    internal static class EnumHelpers
+    {
+        private static readonly ConcurrentDictionary<BclType, EnumConverter> s_converters =
+            new ConcurrentDictionary<BclType, EnumConverter>();
+
+        private static EnumConverter GetConverter(BclType type) =>
+            s_converters.GetOrAdd(type, t => EnumConverter.ForType(type));
+
+        internal static long EnumToInt64(object value) =>
+            GetConverter(value.GetType()).EnumToInt64(value);
+
+        internal static object Int64ToEnum(long value, BclType targetType) =>
+            GetConverter(targetType).Int64ToEnum(value);
+
+        // This is really ugly in terms of repetition, but reasonably simple. I've failed to come up with anything cleaner :(
```

Comment:
Maybe if you have just one GenericEnumeConverter with a second type parameter, and initializ the dict below as: (not entirely sure that would work though)

```
{ TypeCode.Byte, typeof(GenericEnumConverter<, byte>) },
{ TypeCode.SByte, typeof(GenericEnumConverter<, sbyte>) },
{ TypeCode.Int16, typeof(GenericEnumConverter<, short>) },
```

Reply:
I wondered about that, but I don't think it would work - we need the compiler's knowledge of a conversion between `long` and each of the other integer primitive types.

Reply:
Oh yes, that's true.